geopandas/tools/_random.py

```diff
@@ -0,0 +1,330 @@
+    size : integer, tuple
+        an integer denoting how many points to sample, or a tuple
+        denoting how many points to sample, and how many tines to conduct sampling.
+    batch_size: integer
+        a number denoting how large each round of simulation and checking
+        should be. Should be approximately on the order of the number of points
+        requested to sample. Some complex shapes may be faster to sample if the
+        batch size increases. Only useful for (Multi)Polygon geometries.
+
+    Returns
+    -------
+    shapely.MultiPoint geometry containing the sampled points
+
+    Examples
+    --------
+    >>> from shapely.geometry import box
+    >>> square = box(0,0,1,1)
+    >>> grid(square, size=(8,8), tile='hex')
+    """
+
+    if geom is None:
+        geom = geometry.box(0, 0, 1, 1)
+    if size is None:
+        if spacing is None:
+            size = (10, 10)
+        else:
+            ValueError("Either size or spacing options must be provided.")
+    else:  # only process size if it's provided; otherwise, allow spacing to lead
+        if spacing is not None:
+            raise ValueError(
+                "Either size or spacing options can be provided, not both."
+            )
+        if isinstance(size, int):
+            if geom.type in ("Polygon", "MultiPolygon"):
+                size = (size, size)
+            else:
+                size = (size, 1)
+        try:
+            assert isinstance(size, (tuple, list))
+            assert int(size[0]) == size[0]
+            assert int(size[1]) == size[1]
+            assert len(size) == 2
+        except AssertionError:
+            raise TypeError(
+                "Size must be an integer denoting the size of one side of the grid, "
+                " or a tuple of two integers denoting the grid dimensions."
+            )
+    if tile not in ("square", "hex"):
+        raise ValueError(
+            f'The tile option must be either "square" or "hex". Recieved {tile}.'
+        )
+
+    if geom.type in ("Polygon", "MultiPolygon"):
+        multipoints = _grid_polygon(
+            geom,
+            size=size,
+            spacing=spacing,
+            tile=tile,
+            random_offset=random_offset,
+            random_rotation=random_rotation,
+        )
+
+    elif geom.type in ("LineString", "MultiLineString"):
+        multipoints = _grid_line(
+            geom, size=size, spacing=spacing, tile=tile, random_offset=random_offset
+        )
+    else:
+        # TODO: Should we recurse through geometrycollections?
+        multipoints = geometry.MultiPoint()
+
+    return multipoints
+
+
+def _grid_polygon(
+    geom,
+    size=None,
+    spacing=None,
+    tile="square",
+    random_offset=True,
+    random_rotation=True,
+    **unused_kws,
+):
+    """
+    Sample points from within a polygon according to a given spacing.
+    """
+    pygeos = import_optional_dependency(
```

Comment:
I have a solution passing locally with `shapely==2.0b1`:

```python
if SHAPELY_G_20a1:
    spatial_operator = shapely
else:
    spatial_operator = import_optional_dependency(
        "pygeos", "pygeos is required to randomly sample along LineString geometries"
    )
    geom = spatial_operator.from_shapely(geom)
```

However, it is failing in `shapely==1.8.5` with `TypeError: 'mask' should be GeoDataFrame,...`:

<details open>
<summary>Traceback</summary>
<br>

```
__________________________________________________ test_grid[data31-size31-None-square-123456] __________________________________________________

data = <shapely.geometry.polygon.Polygon object at 0x15919a680>, size = (5, 3), spacing = None, tile = 'square', seed = 123456

    @pytest.mark.parametrize(
        ["data", "size", "spacing", "tile", "seed"],
        [
            (data, size, spacing, tile, seed)
            for data in (multipolygon, polygon, multilinestring, linestring)
            for size in (None, 10, (5, 3))
            for spacing in (None, 500)
            for tile in ("hex", "square")
            for seed in (112112, 123456)
            if ((not ((spacing is not None) and (size is not None))))
        ],
    )
    #@pytest.mark.skipif(
    #    not compat.HAS_PYGEOS, reason="Need pygeos to sample random grids from shapes..."
    #)
    def test_grid(data, size, spacing, tile, seed):
        numpy.random.seed(seed)
>       sample = grid(data, size=size, spacing=spacing, tile=tile)

geopandas/tools/tests/test_sample.py:76: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
geopandas/tools/_random.py:143: in grid
    multipoints = _grid_polygon(
geopandas/tools/_random.py:226: in _grid_polygon
    return GeoSeries(points_from_xy(x=x, y=y)).clip(geom).unary_union
geopandas/geoseries.py:1355: in clip
    return geopandas.clip(self, mask=mask, keep_geom_type=keep_geom_type)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

gdf = 0     POINT (939155.295 173761.481)
1     POINT (939493.477 174057.836)
2     POINT (939472.564 173740.567)
3     POIN...  POINT (940107.101 173698.739)
11    POINT (940086.187 173381.471)
12    POINT (940424.370 173677.826)
dtype: geometry
mask = <pygeos.Geometry POLYGON ((939997.095 173013.579, 939915.999 173002.139, 939...>, keep_geom_type = False

    def clip(gdf, mask, keep_geom_type=False):
        """Clip points, lines, or polygon geometries to the mask extent.
    
        Both layers must be in the same Coordinate Reference System (CRS).
        The ``gdf`` will be clipped to the full extent of the clip object.
    
        If there are multiple polygons in mask, data from ``gdf`` will be
        clipped to the total boundary of all polygons in mask.
    
        If the ``mask`` is list-like with four elements ``(minx, miny, maxx, maxy)``, a
        faster rectangle clipping algorithm will be used. Note that this can lead to
        slightly different results in edge cases, e.g. if a line would be reduced to a
        point, this point might not be returned.
        The geometry is clipped in a fast but possibly dirty way. The output is not
        guaranteed to be valid. No exceptions will be raised for topological errors.
    
        Parameters
        ----------
        gdf : GeoDataFrame or GeoSeries
            Vector layer (point, line, polygon) to be clipped to mask.
        mask : GeoDataFrame, GeoSeries, (Multi)Polygon, list-like
            Polygon vector layer used to clip ``gdf``.
            The mask's geometry is dissolved into one geometric feature
            and intersected with ``gdf``.
            If the mask is list-like with four elements ``(minx, miny, maxx, maxy)``,
            ``clip`` will use a faster rectangle clipping (:meth:`~GeoSeries.clip_by_rect`),
            possibly leading to slightly different results.
        keep_geom_type : boolean, default False
            If True, return only geometries of original type in case of intersection
            resulting in multiple geometry types or GeometryCollections.
            If False, return all resulting geometries (potentially mixed-types).
    
        Returns
        -------
        GeoDataFrame or GeoSeries
             Vector data (points, lines, polygons) from ``gdf`` clipped to
             polygon boundary from mask.
    
        See also
        --------
        GeoDataFrame.clip : equivalent GeoDataFrame method
        GeoSeries.clip : equivalent GeoSeries method
    
        Examples
        --------
        Clip points (global cities) with a polygon (the South American continent):
    
        >>> world = geopandas.read_file(
        ...     geopandas.datasets.get_path('naturalearth_lowres'))
        >>> south_america = world[world['continent'] == "South America"]
        >>> capitals = geopandas.read_file(
        ...     geopandas.datasets.get_path('naturalearth_cities'))
        >>> capitals.shape
        (243, 2)
    
        >>> sa_capitals = geopandas.clip(capitals, south_america)
        >>> sa_capitals.shape
        (15, 2)
        """
        if not isinstance(gdf, (GeoDataFrame, GeoSeries)):
            raise TypeError(
                "'gdf' should be GeoDataFrame or GeoSeries, got {}".format(type(gdf))
            )
    
        mask_is_list_like = _mask_is_list_like_rectangle(mask)
        if (
            not isinstance(mask, (GeoDataFrame, GeoSeries, Polygon, MultiPolygon))
            and not mask_is_list_like
        ):
>           raise TypeError(
                "'mask' should be GeoDataFrame, GeoSeries,"
                f"(Multi)Polygon or list-like, got {type(mask)}"
            )
E           TypeError: 'mask' should be GeoDataFrame, GeoSeries,(Multi)Polygon or list-like, got <class 'pygeos.lib.Geometry'>

geopandas/tools/clip.py:153: TypeError
```

</details>


Before spending time trying to hunt this down, is this the kind of solution we are imagining or am I way off base here?


@ljwolf @martinfleis

Reply:
I think that the ideal solution is to make it work if either pygeos or shapely 2.0 are installed. That means that it should work if pygeos and shapely 1.8 are installed but also if shapely 2.0 is installed and there is no pygeos. The implementation should treat the last option as a default because that is what we will need in GeoPandas 1.0. 

If this duality proves to be difficult, we can potentially require shapely 2.0 to be installed and cast possible pygeos geometries to shapely.

Reply:
So swapping the logic would be desired?:

```python
if not SHAPELY_G_20a1:
    spatial_operator = import_optional_dependency(
        "pygeos", "pygeos is required to randomly sample along LineString geometries"
    )
    geom = spatial_operator.from_shapely(geom)
else:
    spatial_operator = shapely
```


Any idea about the `mask` TypeError I am seeing?

Reply:
It gets more complicated than that due to possible backend conflicts. See the proposed solution below.